File: 18_Bit_Manipulation/05_Missing_Number/0268-missing-number.cpp
/*
Problem: LeetCode 268 - Missing Number

Description:
Given an array nums containing n distinct numbers in the range [0, n],
return the only number in the range that is missing from the array.

Intuition:
We can use the mathematical concept of finding the sum of the first n natural numbers
and subtract the sum of the elements in the given array to find the missing number.

Approach:
1. Initialize a variable 'expectedSum' to store the sum of the first n natural numbers.
2. Iterate through the array 'nums' and calculate the sum of its elements, storing it in the variable 'actualSum'.
3. Calculate the missing number as 'expectedSum - actualSum' and return it.

Time Complexity:
The time complexity is O(n) as we need to iterate through the entire array once to calculate the sum of its elements.

Space Complexity:
The space complexity is O(1) as we use only a constant amount of extra space to store variables.
*/

class Solution {
  public:
    int missingNumber(std::vector<int> &nums) {
        int n = nums.size();
        int expectedSum = n * (n + 1) / 2;
        int actualSum = 0;

        for (int num : nums) {
            actualSum += num;
        }

        return expectedSum - actualSum;
    }
};

File: 18_Bit_Manipulation/05_Missing_Number/0268-missing-number.rs
/*
Problem: LeetCode 268 - Missing Number

Key Idea:
We can solve this by calculating the sum of all numbers from 0 to 'n' using the formula 'n * (n + 1) / 2' and subtracting the sum of the elements in the array from it.

Approach:
1. Initialize a variable 'expected_sum' to 0. This variable will store the sum of all numbers from 0 to 'n'.
2. Initialize a variable 'actual_sum' to 0. This variable will store the sum of the elements in the input array.
3. Iterate through the input array and add each element to 'actual_sum'.
4. Calculate 'expected_sum' using the formula 'n * (n + 1) / 2', where 'n' is the length of the input array.
5. The missing number is 'expected_sum - actual_sum'.
6. Return the missing number.

Time Complexity:
O(n), where 'n' is the length of the input array. We iterate through the array once to calculate 'actual_sum' and 'expected_sum'.

Space Complexity:
O(1), as we use only a constant amount of extra space.
*/

impl Solution {
    pub fn missing_number(nums: Vec<i32>) -> i32 {
        let n = nums.len() as i32;
        let expected_sum = n * (n + 1) / 2;
        let actual_sum: i32 = nums.iter().sum();

        expected_sum - actual_sum
    }
}

/*
impl Solution {
    pub fn missing_number(nums: Vec<i32>) -> i32 {
        (nums.len() * (nums.len() + 1) / 2) as i32 - nums.iter().sum::<i32>()
    }
}
*/
